geps/gep-1651.md

```diff
@@ -0,0 +1,218 @@
+# GEP-1651: GatewayAddress Scope
+
+* Issue: [#1651](https://github.com/kubernetes-sigs/gateway-api/issues/1651)
+* Status: Provisional
+
+(See status definitions [here](overview.md#status).)
+
+## TLDR
+
+Allow users to configure a GatewayAddress so that it is only reachable within
+a specific scope (ie. public/cluster local)
+
+## Goals
+
+- Cluster local gateways can be created in a declarative way `ClusterLocal` and
+  are only accessible within a single cluster
+- Define a `Public` scope for addresses that are accessible to the internet
+- Implementations can define & support their own scopes
+
+## Non-Goals
+
+- Per-request/route scope
+- Not a lightweight service mesh
+
+## Introduction
+
+One of the early feature requests for Knative was the ability to deploy an
+application using Knative's HTTP routing support, but make it only available
+within the cluster. I want to be able to specify both the "internal"
+(service.namespace.svc) and "external" (service.namespace.example.com).
+Gateways using the same GatewayClass on the cluster, but ensure that the
+"internal" service is only reachable within the cluster. This would greatly
+simplify deployment for users over the instructions we have today.
+
+The reason why we require an "internal" Gateway is so we can take advantage of
+layer 7 load balancing for our internal services.
+
+## API
+
+We propose adding a new `reachability` field under the `spec.infrastructure` stanza of a Gateway.
+
+### Predefined Reachability Values
+
+Implementations MAY implement the following values for 'reachability' and MUST abide by
+their defined semantics.
+
+Value | Semantic
+-|-
+`public`|The address is reachability over the public internet
+`cluster`|The address is only reachability within the bounds defined by the [cluster's network model](https://kubernetes.io/docs/concepts/cluster-administration/networking/#how-to-implement-the-kubernetes-network-model)
+
+### Vendor prefixed values
+
+Implementations can define custom 'reachability' values by specifying a vendor prefix followed
+by a slash `/` and a custom name ie. `com.ingress.dev/my-reachability`.
+
+### Default reachability
+The default value of `reachability` is implementation specific. It is RECOMMENDED that
+the default `reachability` remains consistent for Gateways with the same
+`gatewayClassName`.
```

Comment:
Would it not be sufficient to tie reachability to a gatewayClassName? There really aren't that many possible options here, why not just have different Gateway Classes?

Reply:
> Would it not be sufficient to tie reachability to a gatewayClassName?

I think the disconnect here is different implementations provision infrastructure at different levels in the API.

If some infrastructure that affects reachability is provisioned when a GatewayClass is created I would recommend tying reachability to a gatewayClass. But I don't think it's sufficient because it is ignoring implementations that have the flexibility to provision infrastructure when a Gateway is created.

Reply:
Are we talking about "reachability" or "default reachability" here? Line 59 is referring to "default reachability" of a gateway, which IMO should be tied to the gateway class. What's the downside of that?

Reply:
Yeah it's default - so I'm recommending  it can inherit from the gateway class

Reply:
>If some infrastructure that affects reachability is provisioned when a GatewayClass is created I would recommend tying reachability to a gatewayClass. But I don't think it's sufficient because it is ignoring implementations that have the flexibility to provision infrastructure when a Gateway is created.

I think most controllers are provisioning Gateways on demand, it's more a question of where the config lives. It's entirely possible for a single controller to implement multiple GatewayClasses. The important detail in a GatewayClass is the `controllerName`. If most implementations can support 1-2 address scopes, why not just leave that for GatewayClasses? In-cluster implementations could include two GatewayClasses by default, for example foo-cluster and foo-public. In my opinion, this seems like the kind of config that could even make sense as a top level attribute in GatewayClass spec.

Reply:
> In my opinion, this seems like the kind of config that could even make sense as a top level attribute in GatewayClass spec.

Do you have a concrete example?

> I think most controllers are provisioning Gateways on demand

If this is the common case why not allow per-gateway options?

>  If most implementations can support 1-2 address scopes, why not just leave that for GatewayClasses?

I think the argument you made [here](https://github.com/kubernetes-sigs/gateway-api/pull/1653#issuecomment-1451246877) about avoiding a combinatorial explotion of address types applies to this scenario.

`foo-public` and `foo-cluster` are two GatewayClasses surfacing the values of a single attribute `reachability`. Let's say we want to enforce IP address type at the same level then our `gatewayClassName` would

`foo-public-ipv4`
`foo-public-ipv6`
`foo-cluster-ipv4`
`foo-cluster-ipv6`

I would also like to echo @howardjohn's points mentioned here: https://github.com/kubernetes-sigs/gateway-api/pull/1653#issuecomment-1429992160

> I know there is "params" on Gatewayclass but I think this is an anti-pattern for most use cases. This would be akin to putting a cluster scoped "ResourceClass" resource that has cpu/memory requests, and deployments reference that vs inlining it - this isn't a terrible idea, but feels more like a higher level abstraction that a PaaS may offer, while I would expect Kubernetes to be a bit lower level. Similarly, having the ability to configure things at a higher level seems nice for Gateway, but being able to configure them on a per-Gateway basis remains important.
>
> I would also like to point out that the "1-2 Gateways per cluster" is a common case, but not the only one. We expect to have 100s of Gateways in a cluster which all may have bespoke config. Making each of theserequire a GatewayClass is problematic